File: aesp/src/main/java/ut/aesp/model/ProgressReport.java
package ut.aesp.model;

import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import java.time.LocalDate;
import java.time.LocalDateTime;

@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Entity
@Table(name = "progress_reports")
public class ProgressReport extends Auditable {

  @Id
  @GeneratedValue(strategy = GenerationType.IDENTITY)
  private Long id;

  @ManyToOne
  @JoinColumn(name = "learner_id", nullable = false)
  private LearnerProfile learner;

  private LocalDate weekStart;
  private LocalDate weekEnd;
  private Integer totalSessions;
  private Float avgPronunciation;
  private Float avgGrammar;
  private Float avgVocabulary;
  private LocalDateTime generatedAt;

  @Lob
  private String improvementNotes;
}

File: ASEP-FE/src/schemas/ProgressReport.ts
import { z } from "zod";
import { LearnerProfileSchema } from "./LearnerProfile"; 

export const ProgressReportSchema = z.object({
  id: z.number().int().optional(),

  learner: LearnerProfileSchema,

  weekStart: z
    .string()
    .refine((val) => !isNaN(Date.parse(val)), {
      message: "weekStart phải là định dạng ngày hợp lệ (ISO)",
    }),

  weekEnd: z
    .string()
    .refine((val) => !isNaN(Date.parse(val)), {
      message: "weekEnd phải là định dạng ngày hợp lệ (ISO)",
    }),

  totalSessions: z
    .number()
    .int()
    .nonnegative({ message: "totalSessions phải >= 0" }),

  avgPronunciation: z
    .number()
    .min(0, { message: "Phải >= 0" })
    .max(10, { message: "Phải <= 10" })
    .nullable()
    .optional(),

  avgGrammar: z
    .number()
    .min(0, { message: "Phải >= 0" })
    .max(10, { message: "Phải <= 10" })
    .nullable()
    .optional(),

  avgVocabulary: z
    .number()
    .min(0, { message: "Phải >= 0" })
    .max(10, { message: "Phải <= 10" })
    .nullable()
    .optional(),

  generatedAt: z
    .string()
    .refine((val) => !isNaN(Date.parse(val)), {
      message: "generatedAt phải là datetime hợp lệ (ISO)",
    }),

  improvementNotes: z.string().nullable().optional(),
});

export type ProgressReport = z.infer<typeof ProgressReportSchema>;
